[libc][fenv] Refactor x86 fenv implementations to make it work for various fenv_t.

lntue opened the pull request.

lntue committed:
[libc][fenv] Refactor x86 fenv implementations to make it work for various fenv_t.

lntue committed:
Fix get_except and include FE_DENORM in FE_ALL_EXCEPT.

lntue committed:
Fix formatting.

lntue committed:
Fix for windows.

enh-google reviewed (approved):
seems plausible, and -- though it's a lot of code compared to bionic's risc stuff -- it's not much worse than bionic's cisc stuff.

michaelrj-google commented on libc/src/__support/FPUtil/x86_64/fenv_x86_common.h:
```diff
@@ -0,0 +1,259 @@
+namespace internal {
+
+// Default order of floating point exception flags in x87 and mxcsr registers:
+// - Bit 1: Invalid Operations
```
below you call the first bit `0` and here you call the first bit `1`. These should probably be consistent

michaelrj-google reviewed (commented):
I did a quick review, I need to go back and go over the larger diffs. Overall looks fine but the bit changes feel like they could be separated out

lntue committed:
- Address comment.
- Add trap on raise fp except option that will match glibc's behavior.
- Fix bazel build.

lntue committed:
Fix feholdexcept_test.

lntue committed:
Fix getenv_and_setenv_test for Windows.

lntue committed:
Fix x87-only detection.

lntue committed:
Update FModTest.h

lntue committed:
Fix fmod's smoke tests.

lntue committed:
Skip denorm exceptions in fmod* tests.

jyknight commented on libc/config/config.json:
```diff
@@ -130,5 +130,11 @@
     }
+  },
+  "fenv": {
+    "LIBC_CONF_TRAP_ON_RAISE_FP_EXCEPT": {
```
Should this be on by default?

jyknight commented on libc/src/__support/CPP/bit.h:
```diff
@@ -43,16 +53,30 @@ bit_cast(const From &from) {
+bit_copy(const From &from, To &to) {
+  MSAN_UNPOISON(&from, sizeof(From));
+  if constexpr (sizeof(To) == sizeof(From)) {
+    to = bit_cast<To>(from);
```
This case calls `To::operator=`, which seems inconsistent with the name "bit_copy" and the else clause.

lntue replied:
Removed.

jyknight commented on libc/src/__support/FPUtil/x86_64/FEnvImpl.h:
```diff
@@ -9,642 +9,244 @@
-#include "src/__support/macros/sanitizer.h"
+#if (defined(__i386__) && !defined(__SSE__)) ||                                \
+    (defined(_M_IX86_FP) && (_M_IX86_FP == 0))
+// When SSE is not available, we will only touch x87 floating point environment.
```
Why do we have an x87-only mode? Does someone actually want to use llvm-libc on a 32-bit x86 machine that lacks SSE2 (e.g., Pentium 3 or older)?

lntue replied:
I guess just for completeness, and the x87_only part is isolated and simple enough?

jyknight commented on libc/src/__support/FPUtil/x86_64/fenv_x86_common.h:
```diff
@@ -0,0 +1,261 @@
+
+// Exception flags are individual bits in the corresponding registers.
+// So, we just OR the bit values to get the full set of exceptions.
+LIBC_INLINE static uint16_t get_status_value_from_except(int excepts) {
```
This function and others like it seem like a generic adapter from internal to external representation which should work for all architectures -- so shouldn't it be in the code implementing the standard-C API interface, not in x86-specific internal utilities?

lntue replied:
To answer your first question, the first one is yes, there are weird libc's, and unfortunately, an example is our own generated `fenv.h` header.  We will definitely need to revisit our `fenv.h` at some point.  OTOH, a bit flexibility on this part will allow users to switch between different the `fenv.h` headers seamlessly, like from their systems' to LLVM libc headers.

To your second and third point, yes, I'll need to migrate and unify all our internal C fenv usages (which are unfortunately spreading over several places.  And among one of those is to allow the internal functions to simply use the underlying libc's fenv implementation.  There are still reasons to carry our own custom implementations, such as:
- users use full build and only pick math functions but not fenv.h functions in the public package
- C fenv functions might be ignored by the compilers without `#pragma STDC FENV_ACCESS ON` and/or extra compilers' flags, and in general, they pessimize quite a lot of floating point optimizations.  OTOH, from what we've seen so far, inline asm and compiler builtins do quite well and still correct.

jyknight reviewed (commented):
Haven't finished going through it, but sending out my initial thoughts.

lntue commented on libc/src/__support/FPUtil/x86_64/fenv_mxcsr_utils.h:
```diff
@@ -0,0 +1,166 @@
+      [[maybe_unused]] volatile float z = EXCEPTION_INPUTS[idx][0];
+      z /= EXCEPTION_INPUTS[idx][1];
+    } else {
+      // For the remaining exceptions, we floating point multiplications.
```
Done.

lntue commented on libc/src/__support/FPUtil/x86_64/fenv_x86_common.h:
```diff
@@ -0,0 +1,261 @@
+  static constexpr uint16_t ERROR = 0xFFFF;
+};
+
+static constexpr uint16_t MXCSR_ROUNDING_CONTROL_BIT_POSITION = 13;
```
Removed.

lntue committed:
Address comments.

lntue commented on libc/src/__support/FPUtil/x86_64/fenv_x86_common.h:
```diff
@@ -0,0 +1,261 @@
+
+LIBC_INLINE static constexpr bool fenv_exceptions_match_x86() {
+  return (FE_INVALID == ExceptionFlags::INVALID_F) &&
+#ifdef __FE_DENORM
```
It is a `#define` in musl.  Without the `#define`, how can we detect the availability of `__FE_DENORM` enum as in glibc?

lntue commented on libc/src/__support/FPUtil/x86_64/fenv_x86_common.h:
```diff
@@ -0,0 +1,261 @@
+  if constexpr (fenv_exceptions_match_x86()) {
+    return static_cast<uint16_t>(excepts & ExceptionFlags::ALL_F);
+  } else {
+    // We will make use of the fact that exception control bits are single
```
It does get compiled to bit mask, but as you mentioned in other comment, glibc does not define the macro `__FE_DENORM`, and so it got compiled to the same as glibc's `FE_ALL_EXCEPT`, which does not cover all exceptions of x86.  So I put the `if constexpr` clause to cover all x86 floating point exceptions.
